packages/fluentui/react-northstar/src/components/Tree/hooks/useTreeActiveState.tsx

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react';
+import { useAutoControlled } from '@fluentui/react-bindings';
+import { TreeProps } from '../Tree';
+import { useStableProps } from './useStableProps';
+import { FlatTree } from './flattenTree';
+import { findIndex, removeItemAtIndex } from './utils';
+import * as _ from 'lodash';
+
+export interface TreeActiveState {
+  activeItemIds: string[];
+  flatTree: FlatTree;
+  toggleActive: (ids: string[], e: React.SyntheticEvent) => void;
+}
+
+export function useTreeActiveState(
+  props: Pick<TreeProps, 'defaultActiveItemIds' | 'activeItemIds' | 'exclusive'>,
+  flatTree: FlatTree,
+): TreeActiveState {
+  const initialActiveItemIds = React.useMemo(() => {
+    const initalValue = [];
+    Object.keys(flatTree).forEach(key => {
+      if (flatTree[key].expanded) {
+        initalValue.push(key);
+      }
+    });
+    return initalValue;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []); // initialValue only needs to be computed on mount
+
+  const [activeItemIds, setActiveItemIdsState] = useAutoControlled<string[]>({
+    defaultValue: props.defaultActiveItemIds,
+    value: props.activeItemIds,
+    initialValue: initialActiveItemIds,
+  });
+
+  // compute a new flattened tree based on activeItemIds state
+  const updatedFlatTree = React.useMemo(() => {
+    const updatedFlatTree = { ...flatTree };
+    Object.keys(updatedFlatTree).forEach(key => {
+      if (updatedFlatTree[key].expanded) {
+        updatedFlatTree[key].expanded = false;
+      }
+    });
+    activeItemIds.forEach(activeId => {
+      updatedFlatTree[activeId].expanded = true;
+    });
+    return updatedFlatTree;
+  }, [activeItemIds, flatTree]);
```

Comment:
```suggestion
  // We want to set `expanded` value on all items to simplify rendering later
  // There is no sense to recreate a flat tree so we simply mutating it
  const flatTreeItems = React.useMemo(() => {
    const updatedFlatTree = { ...baseFlatTreeItems };

    Object.keys(updatedFlatTree).forEach(key => {
      if (updatedFlatTree[key].expanded) {
        updatedFlatTree[key].expanded = false;
      }
    });

    activeItemIds.forEach(activeId => {
      updatedFlatTree[activeId].expanded = true;
    });

    return updatedFlatTree;
  }, [activeItemIds, baseFlatTreeItems]);
```

Reply:
Shift I saw you added comment.

But the code here, it is not mutating, but recreating.

It is recreating here, because we need to know when to re-calculate the `tobeRenderedItemsProps`.
```ts
// this is from useTree.tsx

 const tobeRenderedItemsProps = React.useMemo(
    () => getTobeRenderedItemsProps(props.items, flatTree, props.renderItemTitle),
    [props.items, props.renderItemTitle, flatTree], // this will never be recalculated in the case of mutating
  )
```

Reply:
Talked about it on the call, we are not deep cloning items, make a comment about that.